2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+to specify to Nix the versions of the Haskell packages you depend on.  For
+instance, if you depend on an older version of `aeson`, you need to make sure
+that Nix doesn't accidentally use a recent version.
+
+In general, there are two ways to fix Haskell package versions.  One way is to
+use the Haskell package set defined in nixpkgs, and overwrite individual
+packages that should be different.  The other way is to completely generate your
+own Nix Haskell package set.
+
+1.  _Overwrite individual entries in the Haskell package set in nixpkgs._
+
+    As explained above, nixpkgs has a built-in Haskell package set.  In
+    general, each package is set to the latest version on Hackage[^8].
+
+    If you need to use an older version of a package, you can easily overwrite
+    packages in the nixpkgs Haskell package set.  For instance, if you need to
+    an older version of `aeson`, you can just specify it in your Nix files.
+
+    If you're working on a small open source package with few dependencies,
+    this is an easy method to implement.
+
+    If you're working on a large proprietary application, this method can be
+    quite a lot of tedious work.  You're effectively performing the work of
+    the `cabal` solver, only manually.
+
+2.  _Generate your own Nix Haskell package set._
+
+    If you're working on a large Haskell project with many dependencies, it is
+    often convenient to generate a Haskell package set for use in Nix.
+
+    There are two main ways people generally generate these Haskell package sets.
+
+    One is by using the output of the `cabal` solver and translating the build
+    plan to a set of Nix derivations.  The other is by using a Stackage LTS
+    package set and translating it to a set of Nix derivations.
+
+    Generating your own Nix Haskell package sets can make sure that you are
+    getting the exact same versions of your Haskell dependencies whether you're
+    building with `nix-build`, plain `cabal`, `cabal` + Nix, plain `stack`, or
+    `stack` + Nix.  This can make Nix easy to use on teams where some people
+    prefer `cabal` (without involving Nix at all), some people prefer `stack`
+    with Nix integration, and some people want to run `nix-build` on CI, etc.
+
+    There are tools for generating a set of Nix derivations from either the
+    output of the `cabal` solver or a Stackage package set.  The two most
+    well-known tools are
+    [`haskell.nix`](https://github.com/input-output-hk/haskell.nix)
+    and [`stack2nix2`](https://github.com/input-output-hk/stack2nix).
+
+    I wrote up a short
+    [comparison](https://discourse.nixos.org/t/backport-ghc-8-6-5-to-19-03/3040/7?u=cdepillabout)
+    of these two tools on the offical
+    [Nix Discourse](https://discourse.nixos.org).
+
+    Generating your own Nix Haskell package set is often overkill if you're
+    only writing a small library, but it works very well if you're on a large
+    team or working on a big project.
```

Comment:
Probably would be nice to tell, why somebody wants to generate their own `nixpkgs` set :slightly_smiling_face: Just a comparison with the approach where you don't do this.